### PR TITLE
[playwright] Phase 0–1: Playwright foundation and full @smoke suite

### PR DESCRIPTION
**Pull request by hhovsepy**
### Describe the change

Playwright e2e foundation plus full Cypress `@smoke` coverage for epic #9712, targeting `epic/playwright-migration`. Cypress remains in place (coexistence).

**Infrastructure (Phase 0)**
- `@playwright/test` + `frontend/playwright.config.ts` with projects for current CI frontend suites
- Page objects: `BasePage`, Overview, Services, Graph, Sidebar, Istio Config, Mesh
- Auth setup via `storageState`: anonymous (local) and OpenShift htpasswd (Jenkins OCP)
- Yarn scripts: install/CI install, smoke/all/test-group, JUnit + blob merge, `--last-failed`
- GHA Playwright smoke on the epic branch (`playwright-smoke` suite)

**Full `@smoke` (Phase 1)**
- All Cypress `@smoke` scenarios ported (~32): about, alert, cookie, help, login, logout, sidebar, services toggles, graph prometheus-disabled, mesh local-kiali, Istio config type + validation filters
- OpenShift-only specs skip when auth strategy is not `openshift`
- Logout uses a fresh session so shared `AUTH_FILE` is not invalidated for later projects

### Steps to test the PR

**Local (anonymous)**
1. Kiali UI at `http://localhost:3001`
2. `cd frontend && yarn install && yarn playwright:install chromium`
3. `yarn playwright:run:smoke`
4. Optional: `--headed` or `yarn playwright:ui --project=smoke`

**Jenkins (OCP)**
- Job: `kiali/test-jobs/kiali-playwright-tests`
- `KIALI_PR=10174`
- Prefer `TEST_SET=playwright:run:smoke` with `TEST_TAGS` empty
- Plain `TEST_TAGS=@smoke` is special-cased to the smoke project (after jenkinsfile update); otherwise leave tags empty

### Automation testing

- GHA: Playwright smoke integration suite on this PR
- Jenkins: smoke green against OCP with OpenShift auth
- JUnit: `playwright-results/combined-report.xml` (blob merge + junit fallback)

### Issue reference

Fixes part of #9712 (Phase 0 infra + Phase 1 full `@smoke`). Remaining: `@core-1` / `@core-2`, ambient/multi-cluster, OSSMC, Cypress cutover.

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: frontend/e2e/global-setup/auth.setup.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ */
+setup('authenticate', async ({ page, request }) => {
+  fs.mkdirSync(path.dirname(AUTH_FILE), { recursive: true });
+
+  const authResponse = await request.get('/api/auth/info');
+  expect(authResponse.ok()).toBeTruthy();
+  const authInfo = (await authResponse.json()) as AuthInfo;
+  const strategy = authInfo.strategy ?? 'anonymous';
+
+  process.env.KIALI_AUTH_STRATEGY = strategy;
```
Comment (ScriptingShrimp) [inline frontend/e2e/global-setup/auth.setup.ts]: **Bug — dead code, won't propagate to test workers.**

Playwright suite workers are separate Node.js processes. `process.env` mutations written in the setup worker are invisible to them. The value is never reliably read: every test that needs the strategy already calls `getAuthStrategy(page)` which hits `/api/auth/info` — that is the actual working mechanism.

Remove this line.

File: frontend/e2e/pages/BasePage.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  }
+
+  async waitForLoad(): Promise<void> {
+    await waitForLoadingComplete(this.page);
+  }
+
+  /**
+   * Click with a short retry when the element is briefly detached / covered
+   * (common with PatternFly overlays and React re-renders).
+   */
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/BasePage.ts]: **Anti-pattern: `robustClick` reimplements Playwright actionability, and does it worse.**

Every `locator.click()` in Playwright already retries until the element is attached, visible, stable, not covered by another element, and enabled. This loop runs for the full `actionTimeout` — already `40_000ms` in `playwright.config.ts`. The outer `while` loop and the 250ms sleep add nothing.

**Why the sleep actively hurts**
- `waitForTimeout(250)` is a hard pause even when the element is ready now
- `locator.click({ timeout: 5_000 })` already retries internally; the outer loop is retry-on-top-of-retry
- The `catch` swallows the real error (e.g. `"element intercepted by <div> at (x,y)"`) and replaces it with `robustClick failed`, hiding what actually went wrong

**"But PatternFly overlays"** — Playwright's `receives events` actionability check handles this. If a PF modal or dropdown is covering the target, `locator.click()` waits until the cover dismisses. No outer loop needed.

**"But React re-renders / detached elements"** — Cypress had stale element reference errors because it resolved DOM nodes eagerly (stored a jQuery handle, DOM changed, handle went stale). Playwright locators are lazy — they re-query the DOM on every action. There is no stale element reference problem in Playwright.

**The one real gap** (element visually actionable but click has no effect because a JS handler isn't attached yet) is not solved by retry either — every retry fails the same way. The correct fix is `page.waitForFunction()` on a state condition, or waiting for a network response that signals the handler is wired.

**Provenance** — this was copied from openshift/console's PF5 migration (the `@see` comment above). Console is migrating a different, older codebase. The pattern doesn't transfer here.

**Fix:** delete `robustClick` and `retryOnError`, replace every call site with plain `locator.click()`. If one specific interaction genuinely needs more than 40s, pass `{ timeout: n }` to that one call. If a PF overlay must dismiss before clicking: `await expect(overlay).not.toBeVisible()` then click.

File: frontend/e2e/pages/BasePage.ts [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        return;
+      } catch (err) {
+        lastError = err;
+        await this.page.waitForTimeout(250);
+      }
+    }
+
+    throw lastError instanceof Error ? lastError : new Error(`robustClick failed: ${String(lastError)}`);
+  }
+
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/BasePage.ts]: Same root issue as `robustClick` above. Generic `try/catch + waitForTimeout(500)` retry loops are a Cypress thought pattern (`cy.wait(n)` + retry). Playwright's built-in retry and auto-wait make this unnecessary. Remove `retryOnError` and replace call sites with direct `await` of the Playwright action.

File: frontend/e2e/pages/OverviewPage.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  async open(): Promise<void> {
+    await gotoConsolePage(this.page, 'overview');
+  }
+
+  async openHelpMenu(): Promise<void> {
+    await this.robustClick(this.getBySel('about-help-button'));
+  }
+
+  async openAbout(): Promise<void> {
+    await this.robustClick(this.page.locator('li[role="none"]').filter({ hasText: 'About' }));
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/OverviewPage.ts]: **Fragile selector copied verbatim from Cypress.**

Cypress original (`cypress/integration/common/kiali_about.ts:8`):
```typescript
cy.get('li[role="none"]').contains('About').click();
```

`role="none"` is a PatternFly presentation node, not a semantic role. Playwright has stronger locators available:
```typescript
await this.page.getByRole('menuitem', { name: 'About' }).click();
```

Same issue at lines 25 and 30 in this file (`expectHelpMenuOptions`, `openHelpMenuItem`). When porting from Cypress, upgrade selectors — don't copy them.

File: frontend/e2e/tests/kiali/kiali_about.spec.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { test, expect } from '../../fixtures/kialiFixtures';
+
+test.describe('Kiali help about', () => {
+  test.beforeEach(async ({ overviewPage }) => {
+    await overviewPage.open();
+  });
+
+  test('Open Kiali about page @smoke @core-caching', async ({ overviewPage, page }) => {
```
Comment (ScriptingShrimp) [inline frontend/e2e/tests/kiali/kiali_about.spec.ts:8]: **Convention note — tags in test titles should be a documented team decision.**

Embedding `@smoke @core-caching` in the title string is a Gherkin convention transplanted into Playwright. It works (the config `grep: /@smoke/` matches against the title) but is unconventional and pollutes report output. This approach is intentional per the migration plan, but it must be explicitly documented as the team convention so future specs follow it consistently rather than by accident.

File: frontend/e2e/tests/kiali/kiali_about.spec.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    await expect(page.locator('[href="https://github.com/kiali"]')).toHaveAttribute('href', 'https://github.com/kiali');
+  });
+
+  test('Verify version information is displayed correctly @smoke @core-caching', async ({ overviewPage, page }) => {
+    await overviewPage.openHelpAndAbout();
+
+    const kialiVersion = page.getByTestId('kiali-version');
+    await expect(kialiVersion).toBeVisible();
+    const versionText = (await kialiVersion.innerText()).trim();
+    expect(versionText).not.toEqual('');
```
Comment (ScriptingShrimp) [inline frontend/e2e/tests/kiali/kiali_about.spec.ts:26]: **Negative assertion list copied from Cypress — won't catch new broken formats.**

Cypress original (`cypress/integration/common/kiali_about.ts:26-44`) has the same list. Checking that a value is not `'undefined'`, not `'unknown'`, not `'null'`, not empty documents nothing about what the value *should* be and will pass silently for any novel broken format.

Replace with a positive assertion:
```typescript
expect(versionText).toMatch(/^\d+\.\d+/);
```

Same pattern for `containerText` below.

File: frontend/e2e/pages/IstioConfigPage.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    await this.page.locator('button#filter_select_type-toggle').click();
+    await this.page
+      .locator('div#filter_select_type button')
+      .filter({ hasText: new RegExp(`^${category}$`) })
+      .click();
+  }
+
+  async expectNoActiveFilters(): Promise<void> {
+    await expect(this.page.locator('#filter-selection > :nth-child(2)')).toBeHidden();
+  }
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/IstioConfigPage.ts:59]: **Positional CSS child selector copied from Cypress — breaks on layout changes.**

Cypress original (`cypress/integration/common/istio_config_validation_filters.ts:38`):
```typescript
cy.get('#filter-selection > :nth-child(2)').contains('Valid')
```

`#filter-selection > :nth-child(2)` encodes a positional assumption about the PF chip container's DOM structure. If PF adds a wrapper element or reorders children, this silently passes or fails on the wrong element.

This selector appears in 4 places in this file (lines 51, 101, 125, 151). Add a `data-test="filter-chip-group"` attribute to the source component, or use `page.getByText('Clear all filters')` directly without the parent selector.

File: frontend/e2e/pages/IstioConfigPage.ts [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    await this.page.locator('#filter-selection > :nth-child(2)').getByText('Clear all filters').click();
+    await responsePromise;
+    await waitForLoadingComplete(this.page);
+  }
+
+  async chooseNTypeFilters(count: number): Promise<void> {
+    await this.selectFilterCategory('Type');
+
+    for (let i = 1; i <= count; i++) {
+      await this.page.locator('input[placeholder="Filter by Type"]').click();
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/IstioConfigPage.ts]: **Positional index selector — breaks when filter list order changes.**

```typescript
await this.page.locator(`[data-test=istio-type-dropdown] > :nth-child(${i})`).click();
```

Selecting by position encodes the current ordering of filters as an implicit contract. Add a new filter type or change alphabetical ordering and the test silently selects the wrong one.

Select by name instead:
```typescript
await this.page.getByRole('option', { name: typeName, exact: true }).click();
```

File: frontend/e2e/pages/IstioConfigPage.ts [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  async clickShowLess(): Promise<void> {
+    await this.page.locator('#filter-selection > :nth-child(2)').getByText('Show Less').click();
+    await waitForLoadingComplete(this.page);
+  }
+
+  async expectValidationDropdownVisible(): Promise<void> {
+    await expect(this.page.locator('button#filter_select_value-toggle')).toBeVisible();
+  }
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/IstioConfigPage.ts:138]: **Dropdown opened and closed once per option — copied from Cypress `optionCheck`.**

Cypress original (`cypress/integration/common/istio_config_validation_filters.ts:12-18`) opened and closed the dropdown for each individual option check. 8 clicks to verify 4 options.

Open once, assert all options visible, close once:
```typescript
async expectAllValidationFilterOptions(): Promise<void> {
  await this.page.locator('button#filter_select_value-toggle').click();
  for (const name of VALIDATION_FILTERS) {
    await expect(this.filterOption(name)).toBeVisible();
  }
  await this.page.locator('button#filter_select_value-toggle').click();
}
```

File: frontend/e2e/pages/IstioConfigPage.ts [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+      await this.page.locator(`[data-test=istio-type-dropdown] > :nth-child(${i})`).click();
+      await waitForLoadingComplete(this.page);
+    }
+  }
+
+  async expectActiveFilterCount(count: number): Promise<void> {
+    const chips = this.page.locator('#filter-selection > :nth-child(2)').locator('button[aria-label^="Close "]');
+    await expect(chips).toHaveCount(count);
+  }
+
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/IstioConfigPage.ts:125]: **PF internal class selector — breaks on PF minor version upgrades.**

`pf-v6-c-label` and `pf-m-overflow` are PatternFly internal class names. They changed from PF5 to PF6 and will change again. This selector will silently stop matching after a PF upgrade.

The correct fix is to add `data-test="filter-show-more"` to the overflow button in the source component (`StatefulFilters` or wherever the chip group renders) and select by `this.getBySel('filter-show-more')`.

**Review by ScriptingShrimp — COMMENTED** [review]
Inline review from `review-pr-10174-playwright.md`. Focus: Cypress patterns that must not propagate into future migration phases. See also the updated `plan-cypress-playwright.md` § *Migration conventions* for the full ruleset.

File: frontend/e2e/pages/OverviewPage.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    for (const option of options) {
+      await expect(this.page.locator('li[role="none"]').filter({ hasText: option })).toBeVisible();
+    }
+  }
+
+  async openHelpMenuItem(title: string): Promise<void> {
+    await this.robustClick(this.page.locator('li[role="none"]').filter({ hasText: title }));
+  }
+
+  async expectModalTitle(title: string): Promise<void> {
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/OverviewPage.ts:34]: **PF internal class selector — `h1.pf-v6-c-modal-box__title` will break on PF upgrades.**

Use the semantic modal heading role instead:
```typescript
await expect(
  this.page.getByRole('heading', { name: title, level: 1 })
).toBeVisible();
```

Or add `data-test="modal-title"` to the source if no semantic role is available.

File: frontend/e2e/global-setup/auth.setup.ts (hunk 1 above)
Comment (hhovsepy) [reply frontend/e2e/global-setup/auth.setup.ts]: Fixed — removed the assignment. Workers read auth from `storageState`, not parent env.

File: frontend/e2e/pages/BasePage.ts (hunk 2 above)
Comment (hhovsepy) [reply frontend/e2e/pages/BasePage.ts]: Removed `robustClick`; call sites now use `locator.click()` with Playwright’s built-in actionability retries.

File: frontend/e2e/pages/BasePage.ts (hunk 3 above)
Comment (hhovsepy) [reply frontend/e2e/pages/BasePage.ts]: Removed `retryOnError` as well; same rationale as `robustClick`.

File: frontend/e2e/pages/OverviewPage.ts (hunk 4 above)
Comment (hhovsepy) [reply frontend/e2e/pages/OverviewPage.ts]: Switched to `getByRole('menuitem', { name: … })` for help menu items.

File: frontend/e2e/tests/kiali/kiali_about.spec.ts (hunk 5 above)
Comment (hhovsepy) [reply frontend/e2e/tests/kiali/kiali_about.spec.ts:8]: Intentional for `playwright.config.ts` project `grep` / Jenkins `PLAYWRIGHT_GREP`. Documented in `frontend/e2e/README.md`.

File: frontend/e2e/tests/kiali/kiali_about.spec.ts (hunk 6 above)
Comment (hhovsepy) [reply frontend/e2e/tests/kiali/kiali_about.spec.ts:26]: Aligned with Cypress: added `not.toBeEmpty()` and `not.toContainText('undefined')`, plus trim checks for `unknown` / `null`.

File: frontend/e2e/pages/IstioConfigPage.ts (hunk 7 above)
Comment (hhovsepy) [reply frontend/e2e/pages/IstioConfigPage.ts:59]: Replaced with `#filter-selection` scoped close-button count and `getByRole('button', { name: 'Clear all filters' })`.

File: frontend/e2e/pages/IstioConfigPage.ts (hunk 8 above)
Comment (hhovsepy) [reply frontend/e2e/pages/IstioConfigPage.ts]: `chooseNTypeFilters` now applies `TYPE_FILTERS.slice(0, n)` by name via `filterOption()`.

File: frontend/e2e/pages/IstioConfigPage.ts (hunk 9 above)
Comment (hhovsepy) [reply frontend/e2e/pages/IstioConfigPage.ts:138]: `expectAllValidationFilterOptions` opens the dropdown once, asserts all options, then closes once.

File: frontend/e2e/pages/IstioConfigPage.ts (hunk 10 above)
Comment (hhovsepy) [reply frontend/e2e/pages/IstioConfigPage.ts:125]: Uses `filterSelection().getByRole('button', { name: /^\+/ })` scoped to `#filter-selection` (no PF overflow class).

File: frontend/e2e/pages/OverviewPage.ts (hunk 11 above)
Comment (hhovsepy) [reply frontend/e2e/pages/OverviewPage.ts:34]: Uses `getByRole('heading', { name, level: 1 })` instead of the PF modal title class.

File: frontend/e2e/pages/IstioConfigPage.ts [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    }
+  }
+
+  async expectActiveFilterCount(count: number): Promise<void> {
+    await expect(this.activeFilterCloseButtons()).toHaveCount(count);
+  }
+
+  async showMoreFilters(): Promise<void> {
+    // PF LabelGroup overflow control (e.g. "+1"); Cypress uses the same class selector in label_check.ts.
+    await this.filterSelection().locator('button.pf-v6-c-label.pf-m-overflow').click();
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/IstioConfigPage.ts:128]: **Plan violation: PF internal class selector**

`pf-v6-c-label` and `pf-m-overflow` are PatternFly implementation details. This class naming changed between PF4 → PF5 → PF6; it can change again. The inline comment acknowledging "Cypress uses the same class selector" is exactly the Cypress-copy antipattern the migration plan prohibits.

---

**Option A — semantic selector (no source change, recommended)**

PF `LabelGroup` renders the overflow button as a real `<button>` with accessible text "+N more". Scoped inside `filterSelection()` there is exactly one such button:

```typescript
async showMoreFilters(): Promise<void> {
  await this.filterSelection().getByRole("button", { name: /\+\d+/ }).click();
}
```

Stable as long as PF keeps the button role and "+N" text pattern — both are part of the component's public API, not its internal class structure.

---

**Option B — `data-test` attribute (requires Kiali source change)**

Patch `StatefulFilters.tsx` to expose a `data-test` on the overflow button, then use `getBySel()`:

```typescript
// StatefulFilters.tsx — add data-test to LabelGroup collapsedTextProps
<LabelGroup
  collapsedTextProps={{ "data-test": "label-overflow-button" }}
  ...
>
```

```typescript
// IstioConfigPage.ts
async showMoreFilters(): Promise<void> {
  await this.filterSelection().getBySel("label-overflow-button").click();
}
```

Maximally resilient (zero dependency on text or classes) but requires a source change and verifying PF `LabelGroup` accepts `collapsedTextProps`.

---

**Recommendation:** Option A. The overflow button's accessible name is user-visible interface, not implementation detail. Option B is the right long-term call if `collapsedTextProps` exists in PF6 LabelGroup — worth checking before scaling this pattern to other pages.

**Review by ScriptingShrimp — COMMENTED** [review]
Architecture review comment — see inline.

File: frontend/e2e/tests/kiali/kiali_about.spec.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+    const containerVersion = page.getByTestId('kiali-container-version');
+    await expect(containerVersion).toBeVisible();
+    await expect(containerVersion).not.toBeEmpty();
+    await expect(containerVersion).not.toContainText('undefined');
+    const containerText = (await containerVersion.innerText()).trim();
+    expect(containerText).not.toEqual('');
+    expect(containerText).not.toEqual('unknown');
+    expect(containerText).not.toEqual('null');
+    expect(containerText.length).toBeGreaterThan(0);
```
Comment (ScriptingShrimp) [inline frontend/e2e/tests/kiali/kiali_about.spec.ts:39]: **Plan violation: negative-list assertions**

These blocks assert version validity by enumerating what it should not be. The plan explicitly prohibits this pattern.

Problems:
- `not.toBeEmpty()`, `not.toContainText('undefined')`, `not.toEqual('')/('unknown')/('null')` pass on any garbage that isn't in the list.
- `innerText()` + raw `expect()` (non-web-first) — loses Playwright's auto-retry; a slow render can cause a false failure.
- 7 assertions where 1 is needed.

Fix — replace both blocks with a single positive web-first assertion each:

```typescript
const kialiVersion = page.getByTestId('kiali-version');
await expect(kialiVersion).toBeVisible();
await expect(kialiVersion).toHaveText(/^\d+\.\d+\.\d+/);

const containerVersion = page.getByTestId('kiali-container-version');
await expect(containerVersion).toBeVisible();
await expect(containerVersion).toHaveText(/^\d+\.\d+\.\d+/);
```

The regex matches `2.13.0` and `2.13.0 (abc1234)` (both formats emitted by `AboutUIModal.tsx`), and fails on empty, `unknown`, `null`, `undefined`, or any non-semver string.

File: frontend/e2e/pages/IstioConfigPage.ts [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+
+export class IstioConfigPage extends BasePage {
+  private filterOption(name: string) {
+    // Exact match avoids Gateway⊂K8sGateway and Valid⊂Not Valid / Not Validated.
+    return this.page.locator('#filter_select_value').getByRole('option', { name, exact: true });
+  }
+
+  private filterSelection(): Locator {
+    return this.page.locator('#filter-selection');
+  }
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/IstioConfigPage.ts:34]: **MEDIUM: ID-based selectors — fragile under refactors**

Five distinct IDs used across this file as selectors:

| Selector | Method(s) |
|---|---|
| `#filter_select_value` | `filterOption`, `expectTypeFilterPhrase` |
| `#filter-selection` | `filterSelection`, `removeActiveFilter` |
| `button#filter_select_type-toggle` | `selectFilterCategory` |
| `div#filter_select_type button` | `selectFilterCategory` |
| `button#filter_select_value-toggle` | `expectValidationDropdownVisible`, `expectAllValidationFilterOptions`, `applyValidationFilter`, `chooseNValidationFilters` |

These are Kiali-owned IDs on PF components (not PF-internal), so a PF upgrade won't break them — but a `StatefulFilters.tsx` refactor will break all filter tests simultaneously. Also, IDs must be unique per page; a second toolbar instance would silently select the wrong one.

**Fix — add `data-test` attributes in `StatefulFilters.tsx`:**

```tsx
// MenuToggle id=filter_select_type-toggle
<MenuToggle id="filter_select_type-toggle" data-test="filter-type-toggle" ...>

// Select id="filter_select_type"  
<Select id="filter_select_type" data-test="filter-type-select" ...>

// all MenuToggle id="filter_select_value-toggle" (3 occurrences across renderInput branches + label-op)
<MenuToggle id="filter_select_value-toggle" data-test="filter-value-toggle" ...>

// all Select id="filter_select_value" (3 occurrences)
<Select id="filter_select_value" data-test="filter-value-select" ...>

// Toolbar id="filter-selection"
<Toolbar id="filter-selection" data-test="filter-toolbar" ...>
```

**Then update `IstioConfigPage.ts`:**

```typescript
private filterOption(name: string) {
  return this.page.getByTestId('filter-value-select').getByRole('option', { name, exact: true });
}

private filterSelection(): Locator {
  return this.page.getByTestId('filter-toolbar');
}

async selectFilterCategory(category: 'Type' | 'Config'): Promise<void> {
  await this.page.getByTestId('filter-type-toggle').click();
  await this.page.getByTestId('filter-type-select').getByRole('option', { name: category, exact: true }).click();
}

// expectValidationDropdownVisible / expectAllValidationFilterOptions / applyValidationFilter / chooseNValidationFilters:
// replace this.page.locator('button#filter_select_value-toggle') → this.page.getByTestId('filter-value-toggle')
```

`selectFilterCategory` also drops the `div#filter_select_type button` + `.filter({ hasText })` chain in favour of `getByRole('option')`, consistent with how `filterOption()` already selects from the value dropdown.

File: frontend/e2e/pages/IstioConfigPage.ts [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+    for (const name of TYPE_FILTERS) {
+      await expect(this.filterOption(name)).toBeVisible();
+    }
+  }
+
+  async applyTypeFilter(typeName: string): Promise<void> {
+    const responsePromise = this.page.waitForResponse(
+      response => response.url().includes('/api/istio/config') && response.request().method() === 'GET'
+    );
+    const input = this.page.locator('input[placeholder="Filter by Type"]');
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/IstioConfigPage.ts:83]: **MEDIUM: Placeholder-based selector — fragile under i18n and PF refactors**

Three occurrences of `locator('input[placeholder="Filter by Type"]\')' across `typeIntoTypeFilter`, `expandTypeFilterDropdown`, and `applyTypeFilter`.

Placeholder text is defined in `FiltersAndSorts.ts` as a plain string and passed through `t()` in `StatefulFilters.tsx`. Any i18n work or placeholder rename silently breaks all three selectors simultaneously.

The `TextInputGroupMain` that renders this input already has `id="typeahead-select-input"`. Fix is to add `data-test` there and use it in the POM.

**`StatefulFilters.tsx` — add `data-test` to the typeahead input:**
```tsx
<TextInputGroupMain
  id=typeahead-select-input
  data-test=filter-type-input
  ...
/>
```

**`IstioConfigPage.ts` — replace all three occurrences:**
```typescript
// before
this.page.locator('input[placeholder="Filter by Type"]')

// after
this.page.getByTestId('filter-type-input')
```

File: frontend/e2e/global-setup/auth.setup.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  } else if (strategy === 'token') {
+    throw new Error(
+      'token auth setup is not implemented yet. Use anonymous or openshift auth, or extend auth.setup.ts.'
+    );
+  } else if (strategy === 'openid') {
+    throw new Error(
+      'openid auth setup is not implemented yet. Use anonymous or openshift auth, or extend auth.setup.ts.'
+    );
+  } else {
+    throw new Error(`Unsupported auth strategy: ${strategy}`);
```
Comment (ScriptingShrimp) [inline frontend/e2e/global-setup/auth.setup.ts:46]: **MEDIUM: Hard throws in setup abort the whole suite — zero JUnit output**

When `auth.setup.ts` throws, Playwright marks the `setup` project as crashed. All 18 dependent projects then skip with "Skipped due to failing setup". The JUnit reporter emits nothing for those projects — Jenkins sees an empty report and the pipeline breaks with no actionable output.

The `throw` path also loses the specific strategy in the error context (e.g. "token" vs. "openid" vs. something completely unknown are three separate branches but identical failure mode).

**Fix — write empty storageState first, then fail via `expect` (not `throw`):**

```typescript
} else {
  // Write empty storageState before failing so dependent projects produce JUnit
  // output (skipped) instead of crashing with zero results, which breaks Jenkins.
  fs.writeFileSync(AUTH_FILE, JSON.stringify({ cookies: [], origins: [] }));
  const implemented = ['anonymous', 'openshift'];
  expect(
    implemented,
    `Auth strategy "${strategy}" is not implemented in auth.setup.ts — add support or switch to: ${implemented.join(', ')}`
  ).toContain(strategy);
  return;
}
```

Result: setup project shows **1 FAILED test** with a clear message; dependent projects show **18 SKIPPED** — Jenkins sees failures and skips instead of empty output, and the error points directly to the unsupported strategy.

File: frontend/e2e/tests/graph/graph_display.spec.ts [hunk 17]
```diff
@@ -0,0 +1,8 @@
+import { test } from '../../fixtures/kialiFixtures';
+
+test.describe('Graph display', () => {
+  test('Graph shows empty state when Prometheus is disabled @smoke @prometheus-disabled', async ({ graphPage }) => {
```
Comment (ScriptingShrimp) [inline frontend/e2e/tests/graph/graph_display.spec.ts:4]: **MEDIUM: `@prometheus-disabled` is a floating tag — no project in `playwright.config.ts` routes on it**

`@prometheus-disabled` in the test title has no matching `grep` project. It does nothing for test selection — the test runs only because `@smoke` matches the smoke project. The floating tag is invisible to CI operators and could be mistaken for a working project selector.

Tags in test titles are only meaningful when a project in `playwright.config.ts` has a matching `grep` regex. Conditions that aren't project selectors should use the Playwright `annotation` API instead.

**Fix:**
```typescript
test('Graph shows empty state when Prometheus is disabled @smoke', {
  annotation: { type: 'condition', description: 'Requires Prometheus to be disabled in the cluster' }
}, async ({ graphPage }) => {
  await graphPage.openWithPrometheusDisabled();
  await graphPage.expectPrometheusDisabledEmptyState();
});
```

`@smoke` stays in the title (smoke project grep picks it up). The cluster condition is recorded as a structured annotation — visible in the HTML report and queryable via `testInfo.annotations`, without polluting the project routing namespace.

File: frontend/e2e/pages/SidebarPage.ts [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+    if (hidden === 'true') {
+      await this.navToggle.click();
+    }
+    await expect(this.sidebar).toBeVisible();
+  }
+
+  async ensureSidebarClosed(): Promise<void> {
+    await this.waitForLoad();
+    await expect(this.sidebar).toBeAttached();
+    const hidden = await this.sidebar.getAttribute('aria-hidden');
```
Comment (ScriptingShrimp) [inline frontend/e2e/pages/SidebarPage.ts:26]: **LOW: Mixed visibility semantics — `aria-hidden` vs `.toBeVisible()` can diverge**

`ensureSidebarOpen` reads `aria-hidden` to decide whether to click (line 16), then asserts `toBeVisible()` (line 20). `ensureSidebarClosed` does the same (line 26 / line 30).

These two checks measure different things:
- `aria-hidden='true'` means the element is hidden from the accessibility tree — PF can set this without changing CSS visibility or `display`.
- `toBeVisible()` checks CSS visibility/display — an element can be `aria-hidden` but still visually rendered, or vice-versa.

If PF ever sets `aria-hidden` before the CSS transition completes (or leaves `aria-hidden` set after re-opening), the guard condition and the assertion will disagree — the click will be skipped but the subsequent `toBeVisible()` will fail.

**Fix — use a single consistent signal for both the guard and the assertion:**
```typescript
async ensureSidebarOpen(): Promise<void> {
  await this.waitForLoad();
  if (await this.sidebar.isHidden()) {
    await this.navToggle.click();
  }
  await expect(this.sidebar).toBeVisible();
}

async ensureSidebarClosed(): Promise<void> {
  await this.waitForLoad();
  if (await this.sidebar.isVisible()) {
    await this.navToggle.click();
  }
  await expect(this.sidebar).not.toBeVisible();
}
```
`isHidden()`/`isVisible()` use the same CSS-visibility check as `toBeVisible()`, so guard and assertion can never disagree.

File: frontend/e2e/utils/namespace.ts [hunk 19]
```diff
@@ -0,0 +1,10 @@
+import type { Page } from '@playwright/test';
+import { waitForLoadingComplete } from './transition';
+
+/** Select a namespace in the Kiali namespace dropdown (PatternFly checkbox list). */
+export const selectNamespace = async (page: Page, namespace: string): Promise<void> => {
+  await page.getByTestId('namespace-dropdown').click();
+  await page.locator(`input[type="checkbox"][value="${namespace}"]`).check();
```
Comment (ScriptingShrimp) [inline frontend/e2e/utils/namespace.ts:7]: **LOW: Attribute-value checkbox selector — fragile if PF changes checkbox value semantics**

`input[type="checkbox"][value="${namespace}"]` relies on PF rendering the namespace name as the checkbox `value` attribute. If PF changes the internal checkbox implementation (e.g. moves to a custom component without a native `value`), this silently matches nothing and `.check()` times out.

**Fix — use a role-based selector instead:**
```typescript
await page.getByRole('checkbox', { name: namespace }).check();
```
`getByRole` matches by accessible name (label text), which is stable across PF internal changes and visible in test traces.

File: frontend/playwright.config.ts [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+ * Projects mirror hack/run-integration-tests.sh suites.
+ * Suite membership is controlled by grep tags in test titles (e.g. `@smoke`).
+ * Overlapping tags use negative lookaheads (e.g. @waypoint vs @waypoint-tracing).
+ */
+export default defineConfig({
+  testDir: './e2e/tests',
+  fullyParallel: true,
+  forbidOnly: isCI,
+  retries: isCI ? 1 : 0,
+  workers: isCI ? 2 : undefined,
```
Comment (ScriptingShrimp) [inline frontend/playwright.config.ts:36]: **LOW: `workers: 2` in CI is undocumented — intent not obvious**

`isCI ? 2 : undefined` caps CI parallelism at 2 workers while leaving local runs at Playwright's default (`cpus / 2`). Without a comment, the next person touching this will either remove the cap (risking resource contention on Jenkins) or leave it without understanding why.

```typescript
// Jenkins shared runners have limited CPU/memory; cap at 2 to avoid OOM and
// flaky timeouts under contention. Local runs use Playwright's cpu/2 default.
workers: isCI ? 2 : undefined,
```

File: frontend/e2e/pages/IstioConfigPage.ts (hunk 12 above)
Comment (ScriptingShrimp) [inline frontend/e2e/pages/IstioConfigPage.ts:128]: **HIGH: PF internal class selector — and the original fix had the wrong regex**

`button.pf-v6-c-label.pf-m-overflow` is a PF-internal class, banned by the plan. The comment acknowledges this is copied from Cypress.

The prior attempt used `getByRole('button', { name: /^\+/ })` which didn't match because PF LabelGroup's default `collapsedText` is `'${remaining} more'` (e.g. `"1 more"`) — there is no `+` prefix. The regex was simply wrong.

**Fix — no source changes needed:**
```typescript
async showMoreFilters(): Promise<void> {
  await this.filterSelection().getByRole('button', { name: /\d+ more/ }).click();
}
```

Verified against PF v6 `LabelGroup.tsx`: overflow renders as `<button>` (via `Label variant="overflow"`) with text content from `collapsedText` defaulting to `"${remaining} more"`.

**Comment by hhovsepy** [comment]
Hi @ScriptingShrimp, thank you for suggestions, as this is a work in progress branch, feel free to send a PR with improvements.
I will merge and continue working on next steps not to be blocked.
